### PR TITLE
[Unicode] support `int(str, base=10)` as cpython does

### LINKED ISSUE #5650: Unable to create integer from unicode string with int()
I am trying to convert a unicode string to an integer like this:

```
import numba

@numba.jit(nopython=True)
def str_to_int(str_date):
	return int(str_date[0:4])

print(str_to_int("2016 foo bar"))
```

and receive the following error:

```Invalid use of Function(<class 'int'>) with argument(s) of type(s): (unicode_type)
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<class 'int'>)
[2] During: typing of call at numba-bug.py (5)


File "numba-bug.py", line 5:
def str_to_int(str_date):
	return int(str_date[0:4])```


I have also tried to replace the `int(str_date[0:4])` with `int(str_date[0]) * 1000 + int(str_date[1]) * 100 + int(str_date[2]) * 10 + int(str_date[3])` but the same error appears.

Is this a real bug or am I using the library wrong? Unfortunately I was unable to find more info on substring handling in the docs. Thank you!

### LINKED ISSUE #5723: Overload float(), int() and str() to do number <-> string conversion
So currently all three of these jitted functions will fail

```python
@njit
def failed_str():
	str(1.2)
@njit
def failed_float():
	float("1.2")
@njit
def failed_int():
	int("1.2")
```
I've tried re-implementing these and overloading them.
Strings -> Numbers
```python
@njit
def str_to_posint(s):
	final_index, result = len(s) - 1, 0
	for i,v in enumerate(s):
		result += (ord(v) - 48) * (10 ** (final_index - i))
	return result

@njit
def str_to_int(s):
	neg = (s[0] == "-")
	if(neg): s = s[1:]
	result = str_to_posint(s)
	return -result if neg else result

@njit
def str_to_float(s):
	neg = (s[0] == "-")
	if(neg): s = s[1:]

	exp,dec_loc,exp_loc = 1,-1, len(s)
	for i,c in enumerate(s):
		if(c == "."): 
			dec_loc=i
		elif(c == "e"):
			exp_loc=i
			exp = str_to_int(s[exp_loc+1:])

	if(dec_loc != -1):
		result = str_to_posint(s[:dec_loc])
		result += float(str_to_posint(s[dec_loc+1:exp_loc])) / np.power(10,(exp_loc-dec_loc-1))
	else:
		result = str_to_posint(s)

	if(exp != 1):
		result *= 10.**exp

	if(neg): result = -result
	return result
``` 
Numbers -> Strings
```python
@overload(str)
def _str(x):
	if isinstance(x,types.Integer):
		def int_to_str(x):
			isneg = int(x < 0.0)
			x = np.abs(x)
			l = 0 
			_x = x
			while _x > 0:
				_x = _x // 10
				l += 1
			s = _empty_string(PY_UNICODE_1BYTE_KIND,l+isneg)
			if(isneg): _set_code_point(s,0,DASH)
			for i in range(l):
				digit = x % 10
				_set_code_point(s,isneg+l-i-1,digit + DIGITS_START)
				x = x // 10
			return s
		return int_to_str
	elif isinstance(x,types.Float):
		def float_to_str(x):
			if(x == np.inf):
				return 'inf'
			elif(x == -np.inf):
				return '-inf'

			isneg = int(x < 0.0)
			x = np.abs(x)
			l1,l2 = 0,-1
			_x = x
			while _x > 0:
				_x = _x // 10
				l1 += 1
			_x = x % 10 
			while _x > 1e-10:
				_x = (_x * 10) % 10
				l2 += 1

			l2 = max(1,l2)
			l = l1+l2+1

			_x = x-np.floor(x)
			s = _empty_string(PY_UNICODE_1BYTE_KIND,l+isneg)
			if(isneg): _set_code_point(s,0,DASH)


			_x = x
			for i in range(l1):
				digit = int(_x % 10)
				# print('a',digit,(isneg+l1)-i-1)
				_set_code_point(s,(isneg+l1)-i-1,digit + DIGITS_START)
				_x = _x // 10

			# if(l2 > 0):
			_set_code_point(s,l1+isneg,DOT)

			_x = x % 10 
			for i in range(l2):
				_x = (_x * 10) % 10
				digit = int(_x)  
				# print('here',i,digit)
				# print('b',digit,(isneg+l1)+i+1)
				_set_code_point(s,(isneg+l1)+i+1,digit + DIGITS_START)
			return s
		return float_to_str
```

But it seems that just letting numba loop lift the native python is pretty consistently faster than anything that I can write.

```python
import timeit
N=1000
def time_ms(f):
		f() #warm start
		return " %0.6f ms" % (1000.0*(timeit.timeit(f, number=N)/float(N)))

@njit
def numba_to_str(x):
	return str(x)

def py_to_str(x):
	return str(x)

from warnings import catch_warnings

with catch_warnings(record=True) as w:
	@jit
	def nb_lotsa_strings():
		for i in range(1000):
			numba_to_str(14.58)

	@jit
	def nb_lotsa_strings_int():
		for i in range(1000):
			numba_to_str(14)

	@jit
	def py_lotsa_strings():
		for i in range(1000):
			py_to_str(14.58)

	@jit
	def py_lotsa_strings_int():
		for i in range(1000):
			py_to_str(14)

	print("nb_to_str",time_ms(nb_lotsa_strings))
	print("nb_int_to_str",time_ms(nb_lotsa_strings_int))
	print("py_to_str",time_ms(py_lotsa_strings))
	print("py_int_to_str",time_ms(py_lotsa_strings_int))
```
```bash
----numba---
nb_to_str  1.287220 ms
nb_int_to_str  0.047568 ms
---python---
py_to_str  0.424743 ms
py_int_to_str  0.246021 ms

```

So my feature request is can str(int/float), float(str) and int(str) just be overloaded to grab the GIL and just call native python? Or might there be a way for these to just call out directly to the cpython implementations?

This is related to #5650 and #4758

### PR DESCRIPTION
**Pull request by dlee992**
I refer to https://docs.python.org/3/reference/lexical_analysis.html#integer-literals, and https://docs.python.org/3/library/functions.html#int. 
But I'm not 100% sure about the consistency between my numba impl. and python impl.

I found I don't have to wait for another PR, after some revisions...

Closes #5650 
And also might close #5723

### CODE REVIEW

**Comment by dlee992** [comment]
Hi, @esc. I feel this one can be marked as ready for the initial review too.

**Comment by dlee992** [comment]
Hi, @stuartarchibald , @guilhermeleobas , this one that handles `int(str, base=10)` can be reviewed too.

**Comment by dlee992** [comment]
Hi, @stuartarchibald @guilhermeleobas , this one is ready for an initial review, I guess.

**Comment by dlee992** [comment]
Yes, you are absolutely right. I don't know why I didn't notice these requirements, even I have already posted the link... I will revise this ASAP.

**Comment by dlee992** [comment]
I have updated this PR according python docs and cpython source code in `listobject.c` and `unicodeobject.c`. And I will add similar testcases borrowed from `test_int.py` in cpython, which contains a huge amount of various tests.

Besides, since Python has a `big-int` concept, while numba only has fixed-bit int, long,  as C language. So I simplify the conversion algorithm used in cpython function `long_from_binary_base` and especially in `long_from_non_binary_base`. I am not sure if the conversion performance is degrading too much, but since numba can only hold a 32- or 64-bit of int/long, I guess the performance won't change too much.

**Comment by dlee992** [comment]
> 1. The argument must be a string, bytes or bytearray instance representing an integer literal in the given base.
> 2. The literal can be preceded by `+` or `-`
> 3. The literal can be surrounded by `whitespace`.
> 4. The default base is 10.
> 5. Valid bases are 0 and 2-36, base 0 means interpret the base from the string.

After full changes, the requirements of 2, 3, 4, 5 are met nicely. About the first one, right now I only support the argument is a `string`, I guess we can defer the supporting for other types later.

About if the string literal overflows after converting to int, I didn't check this restriction, except that I check a condition, `digits > 63`, to prevent the digits in the string is too big to store in a `int64`.

And I add many testcases copied or modified from `test_int.py` in cpython, and remove non `string-to-int` conversion testcases. I think @stuartarchibald  or others interested in this PR can take a second review, thanks.

**Comment by dlee992** [comment]
Hi, @stuartarchibald , @sklam do u have some time to re-review this? Many thanks.

**Comment by dlee992** [comment]
For FYI, 

1. this is the python blog in 2006, describing how they implemented the exact BM algorithm in that period of Python: https://web.archive.org/web/20201107074620/http://effbot.org/zone/stringlib.htm
2. this is about all kinds of string fast matching algorithms: http://www-igm.univ-mlv.fr/~lecroq/string/index.html

**Comment by stuartarchibald** [comment]
@dlee992 I edited the PR description to note that this closes #5650.

**Comment by dlee992** [comment]
Hi, @stuartarchibald and @guilhermeleobas .

This PR is ready for second review after my fully revised changes, if u guys have some time for it. Thanks!

File: docs/source/reference/pysupported.rst
Comment (sklam) [inline docs/source/reference/pysupported.rst:937]: Can you add the implementation limitations here? e.g. only support `str` argument.

File: numba/cpython/unicode.py
Comment (guilhermeleobas) [inline numba/cpython/unicode.py:47]: Can you undo this change, as it is not related to the main patch?

File: numba/tests/test_unicode.py
Comment (guilhermeleobas) [inline numba/tests/test_unicode.py:2931]: For the cases `cfunc(ss)` raises an exception, can you check the exception message to see if it is what you're expecting.

File: numba/core/typing/builtins.py
Comment (guilhermeleobas) [inline numba/core/typing/builtins.py:957]: Can you add a comment here saying that support for `int(str)` is done via `@overload`

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
Hi @dlee992, thanks for working on this.

It looks like you covered everything in the implementation. I tried it locally and it is working as expected. I only have a couple of comments that should be easy to address

File: numba/cpython/unicode.py
Comment (guilhermeleobas) [reply numba/cpython/unicode.py:47]: That's fine. I thought the change was just formatting the lines.

File: numba/tests/test_unicode.py
Comment (dlee992) [reply numba/tests/test_unicode.py:2931]: This one is tricky, I will try to revise it.

**Comment by dlee992** [comment]
Hi, I have revised this patch following your advice. Please re-review this again. And Can you put this one into `0.57` as well?

**Review by guilhermeleobas — DISMISSED** [review]
Hi @dlee992, I'll label your PR as waiting for second review as Stuart was reviewing it before.

**Comment by dlee992** [comment]
Hi, @stuartarchibald , this one can use your thoughtful second review!

**Comment by stuartarchibald** [comment]
> Hi, @stuartarchibald , this one can use your thoughtful second review!

@dlee992 I've made some progress on reviewing this (mostly looked at the implementation), there's a lot of involved code (esp. C pointer arithmetic to Python equivalent) so I'm afraid it's taking a while. I hope to find some more time to complete this next week. Thanks in advance for your understanding and patience.

**Comment by dlee992** [comment]
> > Hi, @stuartarchibald , this one can use your thoughtful second review!
> 
> @dlee992 I've made some progress on reviewing this (mostly looked at the implementation), there's a lot of involved code (esp. C pointer arithmetic to Python equivalent) so I'm afraid it's taking a while. I hope to find some more time to complete this next week. Thanks in advance for your understanding and patience.

Thanks! Looking forward to your further comments! If u have any question about the implementation, please discuss with me at any time.

File: numba/cpython/unicode.py
Comment (stuartarchibald) [inline numba/cpython/unicode.py:2458]: This exception string appears even when used in a case where no `base` is present, which is a little confusing. For example:
```python
@njit
def foo():
    return int('a') # there is no base given
```
It would be good to differentiate between the case of "no base" and "base given but not compatible".

Comment (stuartarchibald) [reply numba/cpython/unicode.py:2458]: Same as above applies for a case like:
```python
@njit
def foo():
    return int('1', -2) # invalid base
```
would be good to have a more explicit message about what the problem is (i.e. base is out of supported range).

File: docs/source/reference/pysupported.rst
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst:938]: Depending on how `int` overflow ends up being resolved in this PR, the docs might need associated limitations noting.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates to this PR @dlee992, it's great to see so many involved unit tests derived from CPython's test suite passing! I've reviewed the changes, most of the outstanding things are small changes, but there are a few questions about resolving TODO's and algorithmic differences that need addressing. Thanks again for working on this!

**Pull request closed without merging**
